core/src/main/java/io/grpc/Metadata.java

```diff
@@ -302,62 +429,75 @@ public boolean containsKey(Key<?> key) {
 
   /**
    * Perform a simple merge of two sets of metadata.
    */
   public void merge(Metadata other) {
-    Preconditions.checkNotNull(other, "other");
-    for (Map.Entry<String, List<MetadataEntry>> keyEntry : other.store.entrySet()) {
-      for (int i = 0; i < keyEntry.getValue().size(); i++) {
-        // Must copy the MetadataEntries since they are mutated. If the two Metadata objects are
-        // used from different threads it would cause thread-safety issues.
-        storeAdd(keyEntry.getKey(), new MetadataEntry(keyEntry.getValue().get(i)));
-      }
+    if (other.size == 0) {
+      return;
     }
+    int remaining = 0;
+    if (size == 0 || (remaining = cap() - len()) < other.len()) {
+      expand(len() + other.len() - remaining);
+    }
+    System.arraycopy(other.namesAndValues, 0, namesAndValues, len(), other.len());
+    System.arraycopy(other.typedValues, 0, typedValues, size, other.size);
+    size += other.size;
   }
 
   /**
    * Merge values for the given set of keys into this set of metadata.
    */
   public void merge(Metadata other, Set<Key<?>> keys) {
     Preconditions.checkNotNull(other, "other");
+    // Use ByteBuffer for equals and hashCode.
+    Map<ByteBuffer, Key<?>> asciiKeys = new HashMap<ByteBuffer, Key<?>>(keys.size());
     for (Key<?> key : keys) {
-      List<MetadataEntry> values = other.store.get(key.name());
-      if (values == null) {
-        continue;
-      }
-      for (int i = 0; i < values.size(); i++) {
-        // Must copy the MetadataEntries since they are mutated. If the two Metadata objects are
-        // used from different threads it would cause thread-safety issues.
-        storeAdd(key.name(), new MetadataEntry(values.get(i)));
+      asciiKeys.put(ByteBuffer.wrap(key.asciiName()), key);
+    }
+    for (int i = 0; i < other.size; i++) {
+      ByteBuffer wrappedNamed = ByteBuffer.wrap(other.name(i));
+      Key<?> key = null;
+      if ((key = asciiKeys.get(wrappedNamed)) != null) {
+        if (len() == 0 || len() == cap()) {
+          int desiredCap = len() + (len() >>> 1) + 4;
+          expand(desiredCap & ~1); // floor to a multiple of 2
+        }
+        name(size, other.name(i));
+        value(size, other.value(i));
+        typedValues[size] = other.typedValues[i] != null
+            ? other.typedValues[i] : key.parseBytes(other.value(i));
+        size++;
       }
     }
   }
 
   @Override
+  @SuppressWarnings("deprecation") // We assert that all valid headers are ascii
   public String toString() {
-    return "Metadata(" + toStringInternal() + ")";
-  }
-
-  private String toStringInternal() {
-    return store.toString();
+    StringBuilder sb = new StringBuilder("Metadata(");
+    for (int i = 0; i < size; i++) {
+      if (i != 0) {
+        sb.append(',');
+      }
+      sb.append(new String(name(i), 0)).append('=');
+      if (typedValues[i] != null) {
+        sb.append(typedValues[i]);
+      } else {
+        sb.append(new String(value(i), 0));
```

Comment:
This is in the toString method right?  Do you care that much about this being correct?  This ctor sets all the high bits in the char array to be 0, which will just show up as random charcaters.  Is having the hex output that valuable?